src/plugin/ipc/file/ptyconnection.cpp

```diff
@@ -285,22 +295,51 @@ PtyConnection::doLocking()
   }
 }
 
+ /*
+  * A. Drain master and slave buffer data into _buf and _slave_buf
+  * B. Refill data again both buffers; since both device are paired
+  *    so to refill master-buffer, write on slave-fd and vice versa.
+  */
 void
 PtyConnection::drain()
 {
   JASSERT(_type != PTY_EXTERNAL);
   saveOptions();
-  if (_type == PTY_MASTER && getpgrp() == tcgetpgrp(_fds[0])) {
-    const int maxCount = 10000;
-    char buf[maxCount];
-    int numRead, numWritten;
-
-    // _fds[0] is master fd
-    numRead = ptmxReadAll(_fds[0], buf, maxCount);
+  JASSERT(tcgetattr(_fds[0], &_termios_p) == 0) (JASSERT_ERRNO);
+  if (_type == PTY_MASTER) {
+    int numRead, numWritten, slaveFd;
+    char slavePtsName[100];
     _ptmxIsPacketMode = ptmxTestPacketMode(_fds[0]);
-    JTRACE("_fds[0] is master(/dev/ptmx)") (_fds[0]) (_ptmxIsPacketMode);
-    numWritten = ptmxWriteAll(_fds[0], buf, _ptmxIsPacketMode);
-    JASSERT(numRead == numWritten) (numRead) (numWritten);
+    _real_ptsname_r(_fds[0], slavePtsName, sizeof(slavePtsName));
+    slaveFd = _real_open(slavePtsName, O_RDWR, 0666);
```

Comment:
- Please add a check for the return value.
 - Why do you want 0666 permissions?